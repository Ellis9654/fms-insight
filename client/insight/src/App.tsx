/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import AppBar from 'material-ui/AppBar/AppBar';
import Tabs, { Tab } from 'material-ui/Tabs';
import Typography from 'material-ui/Typography/Typography';
import Toolbar from 'material-ui/Toolbar/Toolbar';
import Hidden from 'material-ui/Hidden/Hidden';
import { Route, Link } from 'react-router-dom';

import Dashboard from './components/Dashboard';
import CostPerPiece from './components/CostPerPiece';
import Efficiency from './components/Efficiency';
import StationMonitor from './components/StationMonitor';
import LoadingIcon from './components/LoadingIcon';
import { requestLastWeek } from './data/events';
import store from './data/store';

// tslint:disable
const LinkTab: any = Tab as any;
// tslint:enable

const tabsStyle = {
  'align-self': 'flex-end',
  'flex-grow': '1'
};

class App extends React.Component {
  componentDidMount() {
    store.dispatch(requestLastWeek());
  }

  render() {
    const tabs = (full: boolean) => (
      <Tabs fullWidth={full} style={full ? {} : tabsStyle} value={location.pathname} onChange={(e, v) => v}>
        <LinkTab label="Dashboard" component={Link} to="/" value="/"/>
        <LinkTab label="Station Monitor" component={Link} to="/station" value="/station"/>
        <LinkTab label="Cost/Piece" component={Link} to="/cost" value="/cost"/>
        <LinkTab label="Efficiency" component={Link} to="/efficiency" value="/efficiency"/>
      </Tabs>
    );

    const largeAppBar = (
      <AppBar position="static">
        <Toolbar>
          <Typography type="title" style={{'margin-right': '2em'}}>Insight</Typography>
          {tabs(false)}
          <LoadingIcon/>
        </Toolbar>
      </AppBar>
    );

    const smallAppBar = (
      <AppBar position="static">
        <Toolbar>
          <Typography type="title">Insight</Typography>
          <div style={{'flex-grow': '1'}}/>
          <LoadingIcon/>
        </Toolbar>
        {tabs(true)}
      </AppBar>
    );

    return (
      <div className="App">
        <Hidden smDown>
          {largeAppBar}
        </Hidden>
        <Hidden mdUp>
          {smallAppBar}
        </Hidden>
        <main style={{margin: '8px'}}>
          <Route exact path="/" component={Dashboard}/>
          <Route exact path="/station" component={StationMonitor}/>
          <Route exact path="/cost" component={CostPerPiece}/>
          <Route exact path="/efficiency" component={Efficiency}/>
        </main>
      </div>
    );
  }
}

export default App;