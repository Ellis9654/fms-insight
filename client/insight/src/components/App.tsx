/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from "react";
import AppBar from "@material-ui/core/AppBar";
import Tabs from "@material-ui/core/Tabs";
import Tab from "@material-ui/core/Tab";
import Typography from "@material-ui/core/Typography";
import Toolbar from "@material-ui/core/Toolbar";
import Hidden from "@material-ui/core/Hidden";
import HelpOutline from "@material-ui/icons/HelpOutline";
import ExitToApp from "@material-ui/icons/ExitToApp";
import IconButton from "@material-ui/core/IconButton";
import Tooltip from "@material-ui/core/Tooltip";
import CircularProgress from "@material-ui/core/CircularProgress";
import Button from "@material-ui/core/Button";
import Badge from "@material-ui/core/Badge";
import Notifications from "@material-ui/icons/Notifications";
import { User } from "oidc-client";

import Dashboard from "./dashboard/Dashboard";
import CostPerPiece from "./cost-per-piece/CostPerPiece";
import Efficiency from "./efficiency/Efficiency";
import StationMonitor from "./station-monitor/StationMonitor";
import DataExport from "./data-export/DataExport";
import ChooseMode from "./ChooseMode";
import LoadingIcon from "./LoadingIcon";
import * as routes from "../data/routes";
import { Store, connect, mkAC } from "../store/store";
import * as api from "../data/api";
import * as serverSettings from "../data/server-settings";
import logo from "../seedtactics-logo.svg";
import BackupViewer from "./BackupViewer";

const tabsStyle = {
  alignSelf: "flex-end" as "flex-end",
  flexGrow: 1
};

enum TabType {
  Operations_Dashboard,
  Operations_Cycles,

  Quality_Dashboard,
  Quality_Serials,

  Analysis_Efficiency,
  Analysis_CostPerPiece,
  Analysis_DataExport
}

interface HeaderTabsProps {
  readonly demo: boolean;
  readonly full: boolean;
  readonly setRoute: (arg: { ty: TabType; curSt: routes.State }) => void;
  readonly routeState: routes.State;
}

function HeaderTabs(p: HeaderTabsProps) {
  let tabType: TabType = TabType.Operations_Dashboard;
  const loc = p.routeState.current;

  switch (loc) {
    case routes.RouteLocation.Operations_Dashboard:
      tabType = TabType.Operations_Dashboard;
      break;
    case routes.RouteLocation.Operations_Cycles:
      tabType = TabType.Operations_Cycles;
      break;
    case routes.RouteLocation.Quality_Dashboard:
      tabType = TabType.Quality_Dashboard;
      break;
    case routes.RouteLocation.Quality_Serials:
      tabType = TabType.Quality_Serials;
      break;
    case routes.RouteLocation.Analysis_Efficiency:
      tabType = TabType.Analysis_Efficiency;
      break;
    case routes.RouteLocation.Analysis_CostPerPiece:
      tabType = TabType.Analysis_CostPerPiece;
      break;
    case routes.RouteLocation.Analysis_DataExport:
      tabType = TabType.Analysis_DataExport;
      break;
  }

  let tabs: JSX.Element[] = [];
  if (p.demo || loc === routes.RouteLocation.Operations_Dashboard || loc === routes.RouteLocation.Operations_Cycles) {
    tabs.push(<Tab label="Operations" value={TabType.Operations_Dashboard} />);
    tabs.push(<Tab label="Cycles" value={TabType.Operations_Cycles} />);
  }
  if (p.demo || loc === routes.RouteLocation.Quality_Dashboard || loc === routes.RouteLocation.Quality_Serials) {
    tabs.push(<Tab label="Quality" value={TabType.Quality_Dashboard} />);
    tabs.push(<Tab label="Serials" value={TabType.Quality_Serials} />);
  }
  if (
    p.demo ||
    loc === routes.RouteLocation.Analysis_Efficiency ||
    loc === routes.RouteLocation.Analysis_CostPerPiece ||
    loc === routes.RouteLocation.Analysis_DataExport
  ) {
    tabs.push(<Tab label="Efficiency" value={TabType.Analysis_Efficiency} />);
    tabs.push(<Tab label="Cost/Piece" value={TabType.Analysis_CostPerPiece} />);
  }
  if (
    loc === routes.RouteLocation.Analysis_Efficiency ||
    loc === routes.RouteLocation.Analysis_CostPerPiece ||
    loc === routes.RouteLocation.Analysis_DataExport
  ) {
    tabs.push(<Tab label="Data Export" value={TabType.Analysis_DataExport} />);
  }

  return (
    <Tabs
      variant={p.full ? "fullWidth" : "standard"}
      style={p.full ? {} : tabsStyle}
      value={tabType}
      onChange={(e, v) => p.setRoute({ ty: v, curSt: p.routeState })}
    >
      {tabs}
    </Tabs>
  );
}

interface HeaderProps {
  demo: boolean;
  showTabs: boolean;
  showAlarms: boolean;
  showLogout: boolean;

  routeState: routes.State;
  fmsInfo: Readonly<api.IFMSInfo> | null;
  latestVersion: serverSettings.LatestInstaller | null;
  alarms: ReadonlyArray<string> | null;
  setRoute: (arg: { ty: TabType; curSt: routes.State }) => void;
  onLogout: () => void;
}

function Header(p: HeaderProps) {
  let helpUrl = "https://fms-insight.seedtactics.com/docs/client-dashboard.html";
  switch (p.routeState.current) {
    case routes.RouteLocation.Station_LoadMonitor:
    case routes.RouteLocation.Station_InspectionMonitor:
    case routes.RouteLocation.Station_WashMonitor:
    case routes.RouteLocation.Station_Queues:
    case routes.RouteLocation.Station_AllMaterial:
      helpUrl = "https://fms-insight.seedtactics.com/docs/client-station-monitor.html";
      break;
    case routes.RouteLocation.Analysis_Efficiency:
      helpUrl = "https://fms-insight.seedtactics.com/docs/client-efficiency.html";
      break;
    case routes.RouteLocation.Analysis_CostPerPiece:
      helpUrl = "https://fms-insight.seedtactics.com/docs/client-cost-per-piece.html";
      break;
  }

  const alarmTooltip = p.alarms ? p.alarms.join(". ") : "No Alarms";
  const Alarms = () => (
    <Tooltip title={alarmTooltip}>
      <Badge badgeContent={p.alarms ? p.alarms.length : 0}>
        <Notifications color={p.alarms ? "error" : undefined} />
      </Badge>
    </Tooltip>
  );

  const HelpButton = () => (
    <Tooltip title="Help">
      <IconButton aria-label="Help" href={helpUrl} target="_help">
        <HelpOutline />
      </IconButton>
    </Tooltip>
  );

  const LogoutButton = () => (
    <Tooltip title="Logout">
      <IconButton aria-label="Logout" onClick={p.onLogout}>
        <ExitToApp />
      </IconButton>
    </Tooltip>
  );

  let tooltip: JSX.Element | string = "";
  if (p.fmsInfo && p.latestVersion) {
    tooltip = (
      <div>
        {(p.fmsInfo.name || "") + " " + (p.fmsInfo.version || "")}
        <br />
        Latest version {p.latestVersion.version} ({p.latestVersion.date.toDateString()})
      </div>
    );
  } else if (p.fmsInfo) {
    tooltip = (p.fmsInfo.name || "") + " " + (p.fmsInfo.version || "");
  }

  const largeAppBar = (
    <AppBar position="static">
      <Toolbar>
        {p.demo ? (
          <a href="/">
            <img src={logo} alt="Logo" style={{ height: "30px", marginRight: "1em" }} />
          </a>
        ) : (
          <Tooltip title={tooltip}>
            <img src={logo} alt="Logo" style={{ height: "30px", marginRight: "1em" }} />
          </Tooltip>
        )}
        <Typography variant="h6" style={{ marginRight: "2em" }}>
          Insight
        </Typography>
        {p.showTabs ? (
          <HeaderTabs demo={p.demo} full={false} setRoute={p.setRoute} routeState={p.routeState} />
        ) : (
          <div style={{ flexGrow: 1 }} />
        )}
        <LoadingIcon />
        {p.showAlarms ? <Alarms /> : undefined}
        <HelpButton />
        {p.showLogout ? <LogoutButton /> : undefined}
      </Toolbar>
    </AppBar>
  );

  const smallAppBar = (
    <AppBar position="static">
      <Toolbar>
        <Tooltip title={tooltip}>
          <img src="/seedtactics-logo.svg" alt="Logo" style={{ height: "25px", marginRight: "4px" }} />
        </Tooltip>
        <Typography variant="h6">Insight</Typography>
        <div style={{ flexGrow: 1 }} />
        <LoadingIcon />
        {p.showAlarms ? <Alarms /> : undefined}
        <HelpButton />
        {p.showLogout ? <LogoutButton /> : undefined}
      </Toolbar>
      {p.showTabs ? (
        <HeaderTabs demo={p.demo} full={true} setRoute={p.setRoute} routeState={p.routeState} />
      ) : (
        undefined
      )}
    </AppBar>
  );

  return (
    <nav id="navHeader">
      <Hidden smDown>{largeAppBar}</Hidden>
      <Hidden mdUp>{smallAppBar}</Hidden>
    </nav>
  );
}

export interface AppProps {
  demo: boolean;
  backupViewerOnRequestOpenFile?: () => void;
}

interface AppConnectedProps extends AppProps {
  route: routes.State;
  fmsInfo: Readonly<api.IFMSInfo> | null;
  user: User | null;
  latestVersion: serverSettings.LatestInstaller | null;
  alarms: ReadonlyArray<string> | null;
  setRoute: (arg: { ty: TabType; curSt: routes.State }) => void;
  onLogin: () => void;
  onLogout: () => void;
}

class App extends React.PureComponent<AppConnectedProps> {
  render() {
    let page: JSX.Element;
    let showTabs: boolean = true;
    let showAlarms: boolean = true;
    let showLogout: boolean = !!this.props.user;
    if (this.props.backupViewerOnRequestOpenFile) {
      page = <BackupViewer onRequestOpenFile={this.props.backupViewerOnRequestOpenFile} />;
      showTabs = false;
      showAlarms = false;
    } else if (this.props.fmsInfo && (!this.props.fmsInfo.openIDConnectAuthority || this.props.user)) {
      switch (this.props.route.current) {
        case routes.RouteLocation.Station_LoadMonitor:
          page = <StationMonitor monitor_type={routes.StationMonitorType.LoadUnload} />;
          break;
        case routes.RouteLocation.Station_InspectionMonitor:
          page = <StationMonitor monitor_type={routes.StationMonitorType.Inspection} />;
          break;
        case routes.RouteLocation.Station_WashMonitor:
          page = <StationMonitor monitor_type={routes.StationMonitorType.Wash} />;
          break;
        case routes.RouteLocation.Station_Queues:
          page = <StationMonitor monitor_type={routes.StationMonitorType.Queues} />;
          break;
        case routes.RouteLocation.Station_AllMaterial:
          page = <StationMonitor monitor_type={routes.StationMonitorType.AllMaterial} />;
          break;

        case routes.RouteLocation.Analysis_CostPerPiece:
          page = <CostPerPiece />;
          showAlarms = false;
          break;
        case routes.RouteLocation.Analysis_Efficiency:
          page = <Efficiency allowSetType={true} />;
          showAlarms = false;
          break;
        case routes.RouteLocation.Analysis_DataExport:
          page = <DataExport />;
          showAlarms = false;
          break;

        case routes.RouteLocation.Operations_Dashboard:
          page = <Dashboard />;
          break;
        case routes.RouteLocation.Operations_Cycles:
          page = <p>Operations Cycles</p>;
          break;

        case routes.RouteLocation.Quality_Dashboard:
          page = <p>Quality Dashboard</p>;
          showAlarms = false;
          break;
        case routes.RouteLocation.Quality_Serials:
          page = <p>Quality Serials</p>;
          showAlarms = false;
          break;

        case routes.RouteLocation.Tools_Dashboard:
          page = <p>Tools Dashboard</p>;
          break;

        case routes.RouteLocation.ChooseMode:
        default:
          page = <ChooseMode />;
          showAlarms = false;
      }
    } else if (this.props.fmsInfo && this.props.fmsInfo.openIDConnectAuthority) {
      page = (
        <div style={{ textAlign: "center", marginTop: "4em" }}>
          <h3>Please Login</h3>
          <Button variant="contained" color="primary" onClick={this.props.onLogin}>
            Login
          </Button>
        </div>
      );
      showTabs = false;
    } else {
      page = (
        <div style={{ textAlign: "center", marginTop: "4em" }}>
          <CircularProgress />
          <p>Loading</p>
        </div>
      );
      showTabs = false;
    }
    return (
      <div id="App">
        <Header
          routeState={this.props.route}
          fmsInfo={this.props.fmsInfo}
          latestVersion={this.props.latestVersion}
          demo={this.props.demo}
          showTabs={showTabs}
          showAlarms={showAlarms}
          showLogout={showLogout}
          setRoute={this.props.setRoute}
          onLogout={this.props.onLogout}
          alarms={this.props.alarms}
        />
        {page}
      </div>
    );
  }
}

function emptyToNull<T>(s: ReadonlyArray<T> | null): ReadonlyArray<T> | null {
  if (!s || s.length === 0) {
    return null;
  } else {
    return s;
  }
}

export default connect(
  (s: Store) => ({
    route: s.Route,
    fmsInfo: s.ServerSettings.fmsInfo || null,
    user: s.ServerSettings.user || null,
    latestVersion: s.ServerSettings.latestInstaller || null,
    alarms: emptyToNull(s.Current.current_status.alarms)
  }),
  {
    setRoute: ({ ty, curSt }: { ty: TabType; curSt: routes.State }): routes.Action => {
      switch (ty) {
        case TabType.Operations_Dashboard:
          return { type: routes.RouteLocation.Operations_Dashboard };
        case TabType.Operations_Cycles:
          return { type: routes.RouteLocation.Operations_Cycles };
        case TabType.Quality_Dashboard:
          return { type: routes.RouteLocation.Quality_Dashboard };
        case TabType.Quality_Serials:
          return { type: routes.RouteLocation.Quality_Serials };
        case TabType.Analysis_Efficiency:
          return { type: routes.RouteLocation.Analysis_Efficiency };
        case TabType.Analysis_CostPerPiece:
          return { type: routes.RouteLocation.Analysis_CostPerPiece };
        case TabType.Analysis_DataExport:
          return { type: routes.RouteLocation.Analysis_DataExport };
      }
    },
    onLogin: mkAC(serverSettings.ActionType.Login),
    onLogout: mkAC(serverSettings.ActionType.Logout)
  }
)(App);
