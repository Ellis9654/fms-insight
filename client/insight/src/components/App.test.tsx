/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';

function mockComponent(name: string): (props: {[key: string]: object}) => JSX.Element {
  return props => (
    <div data-mock-component={name}>
      {Object.getOwnPropertyNames(props).sort().map(p =>
        <span data-prop={p}>
          {JSON.stringify(props[p], null, 2)}
        </span>
      )}
    </div>
  );
}

jest.mock("./LoadingIcon", () => ({
  default: mockComponent("LoadingIcon")
}));
jest.mock("./cost-per-piece/CostPerPiece", () => ({
  default: mockComponent("CostPerPiece")
}));
jest.mock("./dashboard/Dashboard", () => ({
  default: mockComponent("Dashboard")
}));
jest.mock("./efficiency/Efficiency", () => ({
  default: mockComponent("Efficiency")
}));
jest.mock("./station-monitor/StationMonitor", () => ({
  default: mockComponent("StationMonitor")
}));
jest.mock("./data-export/DataExport", () => ({
  default: mockComponent("DataExport")
}));

import * as rtest from 'react-testing-library';

import { App, AppProps, TabType } from './App';
import * as routes from '../data/routes';

function appProps(current: routes.RouteLocation): AppProps {
  let monitor: routes.StationMonitorType = routes.StationMonitorType.LoadUnload;
  switch (current) {
    case routes.RouteLocation.InspectionMonitor:
      monitor = routes.StationMonitorType.Inspection;
      break;
    case routes.RouteLocation.WashMonitor:
      monitor = routes.StationMonitorType.Wash;
  }
  return {
    route: {
      current,
      station_monitor: monitor,
      selected_load_id: 5,
      load_queues: ["q1", "q1"],
      load_free_material: false,
      standalone_queues: ["r1", "r2"],
      standalone_free_material: true,
    },
    fmsInfo: {
      name: "apptest",
      version: "1.2.3.4",
      requireScanAtWash: false,
      requireWorkorderBeforeAllowWashComplete: false,
    },
    latestVersion: {
      date: new Date(2018, 7, 5, 15, 45, 3),
      version: "9.8.7"
    },
    setRoute: jest.fn(),
  };
}

afterEach(rtest.cleanup);

describe("Application", () => {
  it('renders the app', () => {
    const props = appProps(routes.RouteLocation.Dashboard);
    const result = rtest.render(<App {...props}/>);
    expect(result.container).toMatchSnapshot("dashboard");
  });

  it("switches a tab", () => {
    const props = appProps(routes.RouteLocation.Dashboard);
    const result = rtest.render(<App {...props}/>);

    rtest.fireEvent.click(result.getByText("Cost/Piece"));
    expect(props.setRoute).toHaveBeenCalledWith({ty: TabType.CostPerPiece, curSt: props.route});

    rtest.fireEvent.click(result.getByText("Station Monitor"));
    expect(props.setRoute).toHaveBeenCalledWith({ty: TabType.StationMonitor, curSt: props.route});

    rtest.fireEvent.click(result.getByText("Data Export"));
    expect(props.setRoute).toHaveBeenCalledWith({ty: TabType.DataExport, curSt: props.route});

    rtest.fireEvent.click(result.getByText("Efficiency"));
    expect(props.setRoute).toHaveBeenCalledWith({ty: TabType.Efficiency, curSt: props.route});

    rtest.fireEvent.click(result.getByText("Dashboard"));
    expect(props.setRoute).toHaveBeenCalledWith({ty: TabType.Dashboard, curSt: props.route});
  });

});