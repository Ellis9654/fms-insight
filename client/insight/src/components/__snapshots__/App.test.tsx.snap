// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application renders the app: dashboard 1`] = `
<div>
  <div
    id="App"
  >
    <nav
      id="navHeader"
    >
      <header
        class="MuiPaper-root-9 MuiPaper-elevation4-15 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-7"
      >
        <div
          class="MuiToolbar-root-36 MuiToolbar-regular-38 MuiToolbar-gutters-37"
        >
          <img
            alt="Logo"
            style="height: 30px; margin-right: 1em;"
          />
          <h2
            class="MuiTypography-root-47 MuiTypography-title-53"
            style="margin-right: 2em;"
          >
            Insight
          </h2>
          <div
            class="MuiTabs-root-72"
          >
            <div
              class="MuiTabs-flexContainer-73"
            >
              <div
                class="MuiTabs-scroller-75 MuiTabs-fixed-76"
                role="tablist"
                style="margin-bottom: 0px;"
              >
                <div
                  class="MuiTabs-flexContainer-73"
                >
                  <button
                    aria-selected="true"
                    class="MuiButtonBase-root-93 MuiTab-root-81 MuiTab-textColorInherit-83 MuiTab-selected-86"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper-89"
                    >
                      <span
                        class="MuiTab-labelContainer-90"
                      >
                        <span
                          class="MuiTab-label-91"
                        >
                          Dashboard
                        </span>
                      </span>
                    </span>
                    <span
                      class="MuiTouchRipple-root-99"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root-93 MuiTab-root-81 MuiTab-textColorInherit-83"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper-89"
                    >
                      <span
                        class="MuiTab-labelContainer-90"
                      >
                        <span
                          class="MuiTab-label-91"
                        >
                          Station Monitor
                        </span>
                      </span>
                    </span>
                    <span
                      class="MuiTouchRipple-root-99"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root-93 MuiTab-root-81 MuiTab-textColorInherit-83"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper-89"
                    >
                      <span
                        class="MuiTab-labelContainer-90"
                      >
                        <span
                          class="MuiTab-label-91"
                        >
                          Efficiency
                        </span>
                      </span>
                    </span>
                    <span
                      class="MuiTouchRipple-root-99"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root-93 MuiTab-root-81 MuiTab-textColorInherit-83"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper-89"
                    >
                      <span
                        class="MuiTab-labelContainer-90"
                      >
                        <span
                          class="MuiTab-label-91"
                        >
                          Cost/Piece
                        </span>
                      </span>
                    </span>
                    <span
                      class="MuiTouchRipple-root-99"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root-93 MuiTab-root-81 MuiTab-textColorInherit-83"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper-89"
                    >
                      <span
                        class="MuiTab-labelContainer-90"
                      >
                        <span
                          class="MuiTab-label-91"
                        >
                          Data Export
                        </span>
                      </span>
                    </span>
                    <span
                      class="MuiTouchRipple-root-99"
                    />
                  </button>
                </div>
                <span
                  class="TabIndicator-root-96 TabIndicator-colorSecondary-98 MuiTabs-indicator-80"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
          </div>
          <div
            data-mock-component="LoadingIcon"
          />
          <a
            aria-label="Help"
            class="MuiButtonBase-root-93 MuiIconButton-root-106"
            href="https://fms-insight.seedtactics.com/docs/client-dashboard.html"
            role="button"
            tabindex="0"
            target="_help"
            title="Help"
          >
            <span
              class="MuiIconButton-label-111"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-112"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-99"
            />
          </a>
        </div>
      </header>
    </nav>
    <div
      data-mock-component="Dashboard"
    />
  </div>
</div>
`;
